README
ReVIEW(執筆支援ツール)
=====================

このツールは青木さんが公開されている ReVIEW を私(nari)がちょっとだけ拡
張したものです。

オリジナルの ReVIEW リポジトリ
http://i.loveruby.net/svn/public/review/

青木さんによる紹介記事
http://www.loveruby.net/d/20080623.html

私が拡張した分のドキュメントは次の場所に記述しています。
doc/format_extend.txt


ChangeLog
Sun Jul 19 13:14:47 2009  Minero Aoki  <aamine@loveruby.net>

	* lib/review/tocprinter.rb: use Chapter#number to get chapter
	  number, instead of generating it by myself.  Preface and
	  postscript does not have chapter number.

Wed Jul  1 21:03:07 2009  Kenshi Muto  <kmuto@debian.org>

	* lib/review/topbuilder.rb: Add singlechoice and multichoice.

Tue Jun 23 15:11:34 2009  Kenshi Muto  <kmuto@debian.org>

	* lib/review/idgxmlbuilder.rb: escape HTML of keyword alternative.

Mon Jun 22 08:44:38 2009  Kenshi Muto  <kmuto@debian.org>

	* lib/review/index.rb: warn duplicate id.

Sat Jun 13 16:11:14 2009  Minero Aoki  <aamine@loveruby.net>

	* lib/review/book.rb (part): review-index -p causes nil access.

Mon Jun  1 16:01:21 2009  Kenshi Muto  <kmuto@debian.org>

	* lib/review/book.rb: Add basedir path when it calls
	  ImageIndex.parse. This change may destroy if users define
	  specific absolete path for image_dir.

Thu May 28 17:21:27 2009  Kenshi Muto  <kmuto@debian.org>

	* lib/review/book.rb: Add env method into Chapter class to
	  support a backward compatibility for @<chap> tag.

Wed May 27 12:59:20 2009  Kenshi Muto  <kmuto@debian.org>

	* Merged from kmuto branch.
	* lib/review/idgxmlbuilder.rb: improve handling empty cells.
	* lib/review/topbuilder.rb: sync styles of topbuilder.rb with
	  idgxmlbuilder.rb.
	* lib/review/idgxmlbuilder.rb: warn instead of error when
	  table size exceeds the limit.
	* lib/review/idgxmlbuilder.rb: allow null table row.
	* lib/review/builder.rb, lib/review/idgxmlbuilder.rb,
	  lib/review/index.rb, lib/review/topbuilder.rb,
	  lib/review/compiler.rb, doc/format_idg.txt:
	  support null id/caption table.
	* lib/review/idgxmlbuilder.rb: maru op supports A-Za-z also.
	* lib/review/idgxmlbuilder.rb: new block //link.
	* doc/format_idg.txt: mention about //link.
	* lib/review/idgxmlbuilder.rb: fix table cell width count.
	* lib/review/idgxmlbuilder.rb: use <span> for the rerefences of
	  image, table, and list.
        * lib/review/idgxmlbuilder.rb, doc/format_idg.txt: new single
	  command //tsize.
	* bin/review-compile, lib/review/idgxmlbuilder.rb: add -nolfinxml
	  option to remove all LFs from XML elements.
	* lib/review/compiler.rb, lib/review/idgxmlbuilder.rb:
	  Use \t prefix of <p> as indented list level.
	* lib/review/idgxmlbuilder.rb: remove \n from <quote>.
	* lib/review/idgxmlbuilder.rb: new column style [world] and
	  [hood].
	* lib/review/idgxmlbuilder.rb: new block //memo, //tip, //info,
	  //planning, //best, //security and //caution.
	* doc/format_idg.txt: mention about added tags.
	* lib/review/idgxmlbuilder.rb: do escape html for
	  some captions.
	* lib/review/idgxmlbuilder.rb, lib/review/topbuilder.rb:
	  new op @<balloon>.
	* doc/format_idg.txt: mention about @<balloon>.
	* lib/review/idgxmlbuilder.rb, bin/review-compile:
	  add --table option to specify the width of tables and
	  make tables InDesign table objects.
	* lib/review/topbuilder.rb: fix column_begin to accept
	  label parameter.
	* lib/review/idgxmlbuilder.rb: new op @<u>, @<icon>, @<ref>,
	  @<pageref>.
	* lib/review/topbuilder.rb: ditto.
	* lib/review/idgxmlbuilder.rb: use <caption> tag instead of
	  <title> in quotedlist and box blocks.
	  move an image header to the bottom of an image.
	* lib/review/idgxmlbuilder.rb, lib/review/topbuilder.rb,
	  lib/review/compiler.rb: emlist and cmd block may have a
	  caption.
	* lib/review/idgxmlbuilder.rb, lib/review/topbuilder.rb,
	  lib/review/compiler.rb, lib/review/htmlbuilder.rb:
	  hand over the number value of <ol> to ol_item.
	* lib/review/idgxmlbuilder.rb: use \A for finding a head instead
	  of ^.
	* lib/review/latexbuilder.rb, lib/review/idgxmlbuilder.rb,
	  lib/review/topbuilder.rb, lib/review/compiler.rb,
	  lib/review/htmlbuilder.rb: add {label} option into headers.
	  label may be reffered by @<ref> op.
	* lib/review/topbuilder.rb: sync ops and blocks with
	  idgxmlbuilder.
	* lib/review/idgxmlbuilder.rb: box caption is optional.
	* doc/format_idg.txt: mention about return, //label and
	  @<labelref>.
	* lib/review/idgxmlbuilder.rb: new inline op @<labelref>, @<tti>.
	  new block //label.
	* lib/review/idgxmlbuilder.rb: call compile_inline instead of
	  escape_html for <footnote> and <ruby>.
	* lib/review/idgxmlbuilder.rb: new block //lead.
	* doc/format_idg.txt: mention about //point.
	* lib/review/idgxmlbuilder.rb: new block //point.
	* doc/format_idg.txt: mention about nonum and circle option.
	* lib/review/idgxmlbuilder.rb: new header option [circle].
	* doc/format_idg.txt: mention about //important, //notice,
	  @<maru>, @<keytop>.
	* lib/review/idgxmlbuilder.rb: new inine op @<keytop>, @<maru>.
	* lib/review/idgxmlbuilder.rb: escape index strings of @<ttbold>
	  before embedding.
	* lib/review/book.rb: increase maximum section level to 5.
	* lib/review/idgxmlbuilder.rb: new inline op @<raw>.
	  //notice may have a caption.
	* lib/review/idgxmlbuilder.rb: //note and //important may have
	  a caption.
	* lib/review/idgxmlbuilder.rb: embed index information to
	  @<ttbold>.
	* doc/format_idg.txt: mention about //important and //expert.
	* lib/review/idgxmlbuilder.rb: new block //expert.
	* lib/review/idgxmlbuilder.rb, lib/review/compiler.rb: support
	  choice list. "space + white box" becomes multiple choice list.
	  "space + white circle" becomes single choice list.
	* lib/review/idgxmlbuilder.rb, bin/review-compile: add
	  --level option. idgxmlbuilder sets a number to the headers up
	  to specified value.
	* doc/format_idg.txt: mention about //important, //notice, //term,
	  //reference.
	* lib/review/idgxmlbuilder.rb: new block //important, //notice,
	  //reference, and //term.
	* lib/review/idgxmlbuilder.rb: surround a table by <tbody> tag.
	* lib/review/idgxmlbuilder.rb: no need to convert &quot, &lt,
	  and &gt.
	* doc/format_idg.txt: new documentation for InDesign builder.
	* lib/review/idgxmlbuilder.rb: output table without doing
	  compile_inline. compile_inline caused a duplicated character
	  escape.

lib/review/htmlbuilder.rb
# $Id: htmlbuilder.rb 4268 2009-05-27 04:17:08Z kmuto $
#
# Copyright (c) 2002-2007 Minero Aoki
#               2008-2009 Minero Aoki, Kenshi Muto
#
# This program is free software.
# You can distribute or modify this program under the terms of
# the GNU LGPL, Lesser General Public License version 2.1.
#

require 'review/builder'
require 'review/htmlutils'
require 'review/htmllayout'
require 'review/textutils'

module ReVIEW

  class HTMLBuilder < Builder

    include TextUtils
    include HTMLUtils

    def extname
      '.html'
    end

    def builder_init(no_error = false)
      @no_error = no_error
    end
    private :builder_init

    def builder_init_file
      @warns = []
      @errors = []
    end
    private :builder_init_file

    def result
      layout_file = File.join(@book.basedir, "layouts", "layout.erb")
      if File.exists?(layout_file)
        messages() +
          HTMLLayout.new(@output.string, @chapter.title, layout_file).result
      else
        messages() + @output.string
      end
    end

    def warn(msg)
      if @no_error
        @warns.push [@location.filename, @location.lineno, msg]
        puts "----WARNING: #{escape_html(msg)}----"
      else
        $stderr.puts "#{@location}: warning: #{msg}"
      end
    end

    def error(msg)
      if @no_error
        @errors.push [@location.filename, @location.lineno, msg]
        puts "----ERROR: #{escape_html(msg)}----"
      else
        $stderr.puts "#{@location}: error: #{msg}"
      end
    end

    def messages
      error_messages() + warning_messages()
    end

    def error_messages
      return '' if @errors.empty?
      "<h2>Syntax Errors</h2>\n" +
      "<ul>\n" +
      @errors.map {|file, line, msg|
        "<li>#{escape_html(file)}:#{line}: #{escape_html(msg.to_s)}</li>\n"
      }.join('') +
      "</ul>\n"
    end

    def warning_messages
      return '' if @warns.empty?
      "<h2>Warnings</h2>\n" +
      "<ul>\n" +
      @warns.map {|file, line, msg|
        "<li>#{escape_html(file)}:#{line}: #{escape_html(msg)}</li>\n"
      }.join('') +
      "</ul>\n"
    end

    def headline(level, label, caption)
      puts '' if level > 1
      if label.nil?
        puts "<h#{level}>#{caption}</h#{level}>"
      else
        puts "<h#{level} id='#{label}'>#{caption}</h#{level}>"
      end
    end

    def column_begin(level, label, caption)
      puts "<div class='column'>"
      headline(level, label, caption)   # FIXME
    end

    def column_end(level)
      puts "</div>"
    end

    def ul_begin
      puts '<ul>'
    end

    def ul_item(lines)
      puts "<li>#{lines.join("\n")}</li>"
    end

    def ul_end
      puts '</ul>'
    end

    def ol_begin
      puts '<ol>'
    end

    def ol_item(lines, num)
      puts "<li>#{lines.join("\n")}</li>"
    end

    def ol_end
      puts '</ol>'
    end

    def dl_begin
      puts '<dl>'
    end

    def dt(line)
      puts "<dt>#{line}</dt>"
    end

    def dd(lines)
      puts "<dd>#{lines.join("\n")}</dd>"
    end

    def dl_end
      puts '</dl>'
    end

    def paragraph(lines)
      # NOTE: 通常の文章に改行が含まれると変な空白スペースが生まれるため。
      puts "<p>#{lines.join("")}</p>"
    end

    def read(lines)
      puts %Q[<p class="lead">\n#{lines.join("\n")}\n</p>]
    end

    def list_header(id, caption)
      puts %Q[<p class="toplabel">リスト#{@chapter.list(id).number}: #{escape_html(caption)}</p>]
    end

    def list_body(lines)
      puts '<div class="caption-code">'
      puts '<pre class="list">'
      lines.each do |line|
        puts detab(line)
      end
      puts '</pre>'
      puts '</div>'
    end

    def source_header(caption)
      puts %Q[<p class="toplabel">▼#{escape_html(caption)}</p>]
    end

    def source_body(lines)
      puts '<div class="caption-code">'
      puts '<pre class="source">'
      lines.each do |line|
        puts detab(line)
      end
      puts '</pre>'
      puts '</div>'
    end

    def listnum_body(lines)
      puts '<div class="code">'
      puts '<pre class="list">'
      lines.each_with_index do |line, i|
        puts detab((i+1).to_s.rjust(2) + ": " + line)
      end
      puts '</pre>'
      puts '</div>'
    end

    def emlist(lines)
      puts '<div class="code">'
      puts '<pre class="emlist">'
      lines.each do |line|
        puts detab(line)
      end
      puts '</pre>'
      puts '</div>'
    end

    def emlistnum(lines)
      puts '<div class="code">'
      puts '<pre class="emlist">'
      lines.each_with_index do |line, i|
        puts detab((i+1).to_s.rjust(2) + ": " + line)
      end
      puts '</pre>'
      puts '</div>'
    end

    def cmd(lines)
      puts '<div class="code">'
      puts '<pre class="cmd">'
      lines.each do |line|
        puts detab(line)
      end
      puts '</pre>'
      puts '</div>'
    end

    def quotedlist(lines, css_class)
      puts %Q[<blockquote><pre class="#{css_class}">]
      lines.each do |line|
        puts detab(line)
      end
      puts '</pre></blockquote>'
    end
    private :quotedlist

    def quote(lines)
      puts "<blockquote>#{lines.join("\n")}</blockquote>"
    end

    def image_image(id, metric, caption)
      puts %Q[<p class="image">]
      puts %Q[<img src="#{@chapter.image(id).path}" alt="(#{escape_html(caption)})">]
      puts %Q[</p>]
      image_header id, caption
    end

    def image_dummy(id, caption, lines)
      puts %Q[<pre class="dummyimage">]
      lines.each do |line|
        puts detab(line)
      end
      puts %Q[</pre>]
      image_header id, caption
    end

    def image_header(id, caption)
      puts %Q[<p class="botlabel">]
      puts %Q[図#{@chapter.image(id).number}: #{escape_html(caption)}]
      puts %Q[</p>]
    end

    def table_header(id, caption)
      puts %Q[<p class="toplabel">表#{@chapter.table(id).number}: #{escape_html(caption)}</p>]
    end

    def table_begin(ncols)
      puts '<table>'
    end

    def tr(rows)
      puts "<tr>#{rows.join('')}</tr>"
    end

    def th(str)
      "<th>#{str}</th>"
    end

    def td(str)
      "<td>#{str}</td>"
    end
    
    def table_end
      puts '</table>'
    end

    def comment(str)
      puts %Q(<p class="comment">[Comment] #{escape_html(str)}</p>)
    end

    def footnote(id, str)
      puts %Q(<p class="comment"><a name="fn-#{id}">#{escape_html(str)}</a></p>)
    end

    def inline_fn(id)
      %Q(<a href="\#fn-#{id}">*#{@chapter.footnote(id).number}</a>)
    end

    def compile_ruby(base, ruby)
      escape_html(base)   # tmp
    end

    def compile_kw(word, alt)
      '<span class="kw">' +
        if alt
        #then escape_html(word + sprintf(@locale[:parens], alt.strip))
        then escape_html(word + " (#{alt.strip})")
        else escape_html(word)
        end +
      '</span>'
    end

    def inline_i(str)
      %Q(<i>#{escape_html(str)}</i>)
    end

    def inline_b(str)
      %Q(<b>#{escape_html(str)}</b>)
    end

    def inline_ami(str)
      escape_html(str)   # tmp
    end

    def inline_dtp(arg)
      # ignore all
      ''
    end

    def inline_code(str)
      %Q(<span class="inline-code">#{str}</span>)
    end

    def text(str)
      str
    end

    def bibpaper_header(id, caption)
      puts %Q(<a name="bib-#{id}">)
      puts "[#{@chapter.bibpaper(id).number}] #{caption}"
      puts %Q(</a>)
    end

    def bibpaper_bibpaper(id, caption, lines)
      puts %Q(<p>)
      lines.each do |line|
        puts detab(line)
      end
      puts %Q(</p>)
    end

    def inline_bib(id)
      %Q(<a href=".#{@book.bib_file.gsub(/re$/, "html")}\#bib-#{id}">[#{@chapter.bibpaper(id).number}]</a>)
    end

    def nofunc_text(str)
      escape_html(str)
    end

  end

end   # module ReVIEW
